### ISSUE TITLE
MacOS: GLUT_ACTIVE_ALT never fired

### ISSUE THREAD

**Issue by alexey-pelykh** [dcnieho/FreeGLUT#94]
https://github.com/dcnieho/FreeGLUT/blob/349a23dcc1264a76deb79962d1c90462ad0c6f50/freeglut/freeglut/src/x11/fg_main_x11.c#L136

This will never react to these x11 events:

Command key:
```
KeyRelease event, serial 34, synthetic NO, window 0xa00001,
    root 0x111, subw 0x0, time 1552464343, (95,98), root:(95,121),
    state 0x10, keycode 63 (keysym 0xffe7, Meta_L), same_screen YES,
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False
```
Option (and Fn+Option - or Alt):
```
KeyRelease event, serial 34, synthetic NO, window 0xa00001,
    root 0x111, subw 0x0, time 1552464903, (95,98), root:(95,121),
    state 0x2000, keycode 66 (keysym 0xff7e, Mode_switch), same_screen YES,
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False
```

It seems that following next steps are possible:

- Do nothing if it's considered by-design behaviour
- On MacOS, map `0x2000` (which is not even in X11?) to `GLUT_ACTIVE_ALT`
- On MacOS, use Command as Alt
- On MacOS, expose Command otherwise?

**Comment by jtsiomb** [comment]
First of all, this is not the official freeglut repo, and your bug reports here might go unnoticed. The official repo and bug tracker are hosted on sourceforge. Also to my knowledge none of the freeglut maintainers are MacOS users, so it's hard to make any changes blindly if it doesn't come from a 3rd party patch. So if you want this fixed, you'll need to fix it and send us a patch. Before any of that though, let's discuss what the correct behaviour should be.

You posted the events corresponding to command and fn+option, but what about option itself? To my understanding, option is the mac version of the alt key. What event does the X server send when option is pressed?

Also it might be illuminating to compare with how Apple's version of GLUT works. Does it map the option key to GLUT_ACTIVE_ALT, or the command key? It might make sense to make the freeglut behavior compatible with that.

**Comment by alexey-pelykh** [comment]
> What event does the X server send when option is pressed?

Fn does not affect Option:

Option:
```
KeyRelease event, serial 34, synthetic NO, window 0x800001,
    root 0xc1, subw 0x0, time 1192835188, (113,311), root:(113,334),
    state 0x2000, keycode 66 (keysym 0xff7e, Mode_switch), same_screen YES,
    XLookupString gives 0 bytes:
    XFilterEvent returns: False
```

Fn+Option:
```
KeyRelease event, serial 34, synthetic NO, window 0x800001,
    root 0xc1, subw 0x0, time 1192901015, (323,766), root:(323,789),
    state 0x2000, keycode 66 (keysym 0xff7e, Mode_switch), same_screen YES,
    XLookupString gives 0 bytes:
    XFilterEvent returns: False
```

> Does it map the option key to GLUT_ACTIVE_ALT, or the command key?

I'll check, yet my bet that it does not.

**Comment by alexey-pelykh** [comment]
Yep, it does not as far as I can understand. See https://opensource.apple.com/source/X11libs/X11libs-60/mesa/Mesa-7.8.2/src/glut/glx/glut_modifier.c.auto.html

```
  if(__glutModifierMask & Mod1Mask)
    modifiers |= GLUT_ACTIVE_ALT;
```

**Comment by alexey-pelykh** [comment]
Gotcha, for some reason I assumed that GLUT framework sources are the ones from above, at least partially. Will check.

**Comment by jtsiomb** [comment]
Ok I found an old mac mini and I tried it with the GLUT.framework myself: GLUT_ACTIVE_ALT is set by pressing the option key, and command does not affect any modifiers. Now I'll have to try and build freeglut on that mac mini and see what it does. Not sure when I'll get around to it though.

**Comment by jtsiomb** [comment]
Alright, I looked into it some more. Indeed by default freeglut does not trigger `GLUT_ACTIVE_ALT`, but that's just because by default XQuartz is set to *not* treat alt as a modifier, but produce the `Mode_switch` keysym instead. The good news is you can change that by going to the XQuartz preferences and clicking the "Option keys send Alt_L and Alt_R checkbox".

While I was at it, I also added support for the command key. I added a new `GLUT_ACTIVE_SUPER` modifier, which should map to the command key on mac keyboards, and the windows key on regular keyboards. I've also added corresponding special keys `GLUT_KEY_SUPER_L` and `GLUT_KEY_SUPER_R`.

To use these you need to build the latest SVN revision (rev 1875) from sourceforge, and `#include <GL/freeglut.h>`, not `<GL/glut.h>`, to have access to the freeglut API extensions.
